### PR TITLE
TAS: Add scheduling gate to Pods admitted by TAS

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

To add the scheduling pods to the PodTemplate (and thus the Pods) admitted by TAS.

#### Which issue(s) this PR fixes:

Part of #2724 

#### Special notes for your reviewer:

No release note here, I will add it in a single PR which adds the API.

The assignment is done in PR #3256 . The gate will be removed in a follow up PR. This PR can be merged indepdendently of computing the assignment or topology ungater, but all 3 PRs are required for the feature to work e2e.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/podset/podset_test.go [hunk 1]
```diff
@@ -163,10 +167,73 @@ func TestFromAssignment(t *testing.T) {
...
+					"f1l1": "f1v1",
+					"f1l2": "f1v2",
+				},
+				Tolerations: []corev1.Toleration{*toleration1.DeepCopy(), *toleration2.DeepCopy()},
+			},
+		},
 	}
 	for name, tc := range cases {
 		t.Run(name, func(t *testing.T) {
 			ctx := context.TODO()
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/podset/podset_test.go]: sure, done

File: pkg/podset/podset.go [hunk 2]
```diff
@@ -40,24 +42,31 @@ var (
...
-		NodeSelector: make(map[string]string),
-		Count:        ptr.Deref(assignment.Count, defaultCount),
-		Labels:       make(map[string]string),
-		Annotations:  make(map[string]string),
+		Name:            assignment.Name,
+		NodeSelector:    make(map[string]string),
+		Count:           ptr.Deref(assignment.Count, defaultCount),
+		Labels:          make(map[string]string),
+		Annotations:     make(map[string]string),
+		SchedulingGates: make([]corev1.PodSchedulingGate, 0),
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/podset/podset.go]: true, updated

File: pkg/podset/podset_test.go [hunk 3]
```diff
@@ -195,6 +262,9 @@ func TestMergeRestore(t *testing.T) {
 			Value:    "t0v",
 			Effect:   corev1.TaintEffectNoSchedule,
 		}).
+		SchedulingGates(corev1.PodSchedulingGate{
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/podset/podset_test.go]: Yeah, I prefer this pattern, just initially tried to fit-into the existing one.

**Review by alculquicondor — COMMENTED** [review]
/lgtm

File: pkg/podset/podset.go [hunk 4]
```diff
@@ -118,6 +127,12 @@ func (podSetInfo *PodSetInfo) Merge(o PodSetInfo) error {
 			podSetInfo.Tolerations = append(podSetInfo.Tolerations, t)
 		}
 	}
+	// make sure we don't duplicate schedulingGates
+	for _, t := range o.SchedulingGates {
+		if slices.Index(podSetInfo.SchedulingGates, t) == -1 {
```
Comment (tenzen-y) [inline pkg/podset/podset.go:132]: Could you switch from this to `HasGate` lib?
https://github.com/kubernetes-sigs/kueue/blob/4c063056616b1486f699fdb4461903abc8fa0a79/pkg/util/pod/pod.go#L26

File: pkg/podset/podset.go [hunk 5]
```diff
@@ -170,6 +187,10 @@ func RestorePodSpec(meta *metav1.ObjectMeta, spec *corev1.PodSpec, info PodSetIn
 		spec.Tolerations = slices.Clone(info.Tolerations)
 		changed = true
 	}
+	if !slices.Equal(spec.SchedulingGates, info.SchedulingGates) {
```
Comment (tenzen-y) [inline pkg/podset/podset.go:190]: Do we need to respect the order of schedulingGate?
I'm wondering if we can ignore the order differences.

**Pull request merged**